### PR TITLE
adding log rotation in kuma-cp

### PR DESCRIPTION
**Pull request by nikita15p**
### Summary
This PR supports log rotation in kuma-cp

### Full changelog
This will add a capability to perform log rotation for kuma-cp binary if needed

With this change, following is the output of kuma-cp run --help
```
Launch Control Plane.

Usage:
  kuma-cp run [flags]

Flags:
  -c, --config-file string   configuration file
  -h, --help                 help for run

Global Flags:
      --log-level string             log level: one of off|info|debug (default "info")
      --log-max-age int              maximum number of days to retain old log files based on the timestamp encoded in their filename (default 30)
      --log-max-retained-files int   maximum number of old log files to retain (default 1000)
      --log-max-size int             maximum size in megabytes of a log file before it gets rotated (default 100)
      --log-output-path string       path of output rotating file Example: if we set it to /tmp/kuma.log then after the file is rotated we will have /tmp/kuma-2021-06-07T09-15-18.265.log
```

### Issues resolved

Fix https://github.com/kumahq/kuma/issues/2099

### Documentation
NA

### Testing

- [ ] Unit tests
- [ ] E2E tests
- [ ] Manual testing on Universal
- [ ] Manual testing on Kubernetes

### CODE REVIEW

**Comment by tharun208** [comment]
can you update the `kumactl` completions test suite with the updated golden files and also update the kumactl `HELP.md` file using `make docs`

**Comment by nikita15p** [comment]
@tharun208 changes in this PR are related to kuma-cp. HELP.md is specific to kumactl.

File: app/kuma-cp/cmd/root.go
Comment (jakubdyszkiewicz) [inline app/kuma-cp/cmd/root.go:56]: I think we should prefix those options with `log-` so it's
```
cmd.PersistentFlags().StringVar(&args.logLevel, "log-level", kuma_log.InfoLevel.String(), kuma_cmd.UsageOptions("log level", kuma_log.OffLevel, kuma_log.InfoLevel, kuma_log.DebugLevel))
cmd.PersistentFlags().StringVar(&args.outputPath, "log-output-path", args.logOutputPath, "path to the file that will be filled with logs")
cmd.PersistentFlags().IntVar(&args.maxBackups, "log-max-retained-files", 1000, "maximum number of the old log files to retain")
cmd.PersistentFlags().IntVar(&args.maxSize, "log-max-size", 100, "maximum size in megabytes of a log file before it gets rotated")
cmd.PersistentFlags().IntVar(&args.maxAge, "log-max-age", 30, "maximum number of days to retain old log files based on the timestamp encoded in their filename")
```

File: app/kuma-cp/cmd/root.go
Comment (jakubdyszkiewicz) [inline app/kuma-cp/cmd/root.go:53]: so let say I pass `/tmp/kuma.log` what happens when the file gets rotated?

File: pkg/log/logger.go
Comment (jakubdyszkiewicz) [inline pkg/log/logger.go:56]: I'd remove those comments. This file should not know anything about CLI arguments. If we change CLI args there is a high change we forgot to change it here and this comments gets invalid

File: app/kuma-cp/cmd/root.go
Comment (jakubdyszkiewicz) [reply app/kuma-cp/cmd/root.go:53]: Ok, I tested it and the result is
```
❯❯❯ ls -lah /tmp/ | grep kuma
-rw-r--r--   1 jakob  wheel   1.0M Jun  7 11:15 kuma-2021-06-07T09-15-18.265.log
-rw-r--r--   1 jakob  wheel   1.0M Jun  7 11:15 kuma-2021-06-07T09-15-58.193.log
-rw-r--r--   1 jakob  wheel   206K Jun  7 11:16 kuma.log
```

Comment (jakubdyszkiewicz) [inline app/kuma-cp/cmd/root.go:53]: I see that if we set this out then we have no logs in the stdout. @nikita15p @kumahq/kuma-maintainers do you think it's expected behavior?

If we want to stick to it, we should at least print something like `Logs will be stored in path_to_log_file` before we set the logger.

Comment (jakubdyszkiewicz) [reply app/kuma-cp/cmd/root.go:53]: I think in the help I'd write `path of the rotating log file. Example: if we set it to /tmp/kuma.log then after the file is rotated we will have /tmp/kuma-2021-06-07T09-15-18.265.log`

**Review by jakubdyszkiewicz — COMMENTED** [review]
that's awesome!

If we want to have this, this should be then consistent also with `kuma-dp` and `kuma-prometheus-sd`

File: app/kuma-cp/cmd/root.go
Comment (nikita15p) [reply app/kuma-cp/cmd/root.go:53]: This "Logs will be stored in path_to_log_file" can be taken into account. I liked the idea.

File: pkg/log/logger.go
Comment (nikita15p) [reply pkg/log/logger.go:56]: so basically when we do not pass these flags from cli, these values gets applied by default.

File: app/kuma-cp/cmd/root.go
Comment (nikita15p) [reply app/kuma-cp/cmd/root.go]: done

File: app/kuma-cp/cmd/root.go
Comment (jakubdyszkiewicz) [inline app/kuma-cp/cmd/root.go:53]: `path of output rotating file Example...` -> `path to the file that will be filled with logs. Example...`

File: app/kuma-cp/cmd/root.go
Comment (jakubdyszkiewicz) [inline app/kuma-cp/cmd/root.go:54]: `of the old`

File: app/kuma-cp/cmd/root.go
Comment (jakubdyszkiewicz) [reply app/kuma-cp/cmd/root.go]: Ok, let's add it then!

Comment (nikita15p) [reply app/kuma-cp/cmd/root.go]: Added. Please review.

File: pkg/log/logger.go
Comment (jakubdyszkiewicz) [inline pkg/log/logger.go:57]: This is just a constructor for logger, it should not log things.

Please log this just before those lines
```
core.SetLogger(core.NewLoggerWithRotation
```

also don't use `"k8s.io/klog/v2"` it's some random logger from K8S that we don't use. Just use `fmt.Printf` since we don't have our logger before we set it

**Review by jakubdyszkiewicz — CHANGES_REQUESTED** [review]

File: pkg/log/logger.go
Comment (nikita15p) [reply pkg/log/logger.go:57]: Right. Even I was thinking which library to use. Thanks, will make this changes.

File: go.mod
Comment (jakubdyszkiewicz) [inline go.mod:61]: Is this new dependency required? I think we can remove it

File: pkg/plugins/ca/provided/ca_cert_validator.go
Comment (jakubdyszkiewicz) [inline pkg/plugins/ca/provided/ca_cert_validator.go]: I don't think this should be a part of the PR. I will be happy to approve this one as a separate PR once we have a confirmation with E2E testing.

**Review by jakubdyszkiewicz — CHANGES_REQUESTED** [review]

File: pkg/plugins/ca/provided/ca_cert_validator.go
Comment (nikita15p) [reply pkg/plugins/ca/provided/ca_cert_validator.go]: Hi @jakubdyszkiewicz It is not part of PR , if you check the files changed. Also if you see above changes , it is shown as outdated

**Review by jakubdyszkiewicz — APPROVED** [review]

**Comment by jakubdyszkiewicz** [comment]
Thank you for iterating on this PR!

**Comment by tharun208** [comment]
@nikita15p, can we have this change also on the helm chart?

cc @austince

**Comment by nikita15p** [comment]
Hi @tharun208 can you tell me exact usecase. What I found while running kuma with helm, currently all the logs get printed in stdout with helm command. So with that flag we can have logs in file, is that really needed?

**Pull request merged**
